feat(a-msgpack): add Wildcard support

rufman opened the pull request:
- Add support for en/decoding wildcards.
- Rework the encoding logic for extensions. Instead of iterating
through all the extension types before encoding any other objects,
do that last. Add an `identifier` key which allow for the
identification of the extenstion type when encoding, without having
to try to encode with each extenstion type.

rufman committed:
feat(a-msgpack): add Wildcard support

- Add support for en/decoding wildcards.
- Rework the encoding logic for extensions. Instead of iterating
through all the extension types before encoding any other objects,
do that last. Add an `identifier` key which allow for the
identification of the extenstion type when encoding, without having
to try to encode with each extenstion type.

rufman committed:
Merge branch 'trunk' into feature/wildcard

timoha commented on packages/a-msgpack/src/ExtData.ts:
```diff
@@ -1,8 +1,10 @@
  */
 export class ExtData {
-  constructor(readonly type: number, readonly data: Uint8Array) {
+  constructor(readonly type: number, readonly data: Uint8Array | typeof WRITE_ONLY_HEADER) {
```
data.length == 0 wouldn't work because data can be undefined/null? what if you make your have a constant of `Uint8Array` of size 0 to signify no data? Then the argument can only be of the same type and you'd only have to check if payload is zero to not write it as well and use its size in header directly rather than having an if statement.

rufman replied:
Hmm yeah this is a good idea. Let me try that

timoha commented on packages/a-msgpack/src/neat/extensions.ts:
```diff
@@ -19,17 +19,16 @@ import { ExtensionCodec } from '../ExtensionCodec';
-type PointerEncoder = (data: unknown) => Uint8Array | null;
+type PointerEncoder = (data: Pointer) => Uint8Array;
 type PointerDecoder = (buffer: Uint8Array) => Pointer;
+type WildcardEncoder = (data: unknown) => null;
```
why not `Wildcard` instead of `unknown`?

rufman replied:
Yeah this is actually wrong. I'm surprise ts didn't catch this. You shouldn't be able to pass any data to the wildcard encoder

timoha commented on packages/a-msgpack/src/Encoder.ts:
```diff
@@ -359,7 +359,14 @@ export class Encoder {
+  }
+
+  encodeExtensionHeader(ext: ExtData): void {
+    const size = ext.data === WRITE_ONLY_HEADER ? 0 : ext.data.length;
```
why not always use ext.data.length if you were to force somehow not allowing ext.data to be null?

timoha commented on packages/a-msgpack/src/Encoder.ts:
```diff
@@ -359,7 +359,14 @@ export class Encoder {
   encodeExtension(ext: ExtData): void {
-    const size = ext.data.length;
+    this.encodeExtensionHeader(ext);
+    if (ext.data !== WRITE_ONLY_HEADER) {
```
also, why not check for length? or maybe even writeU8a already handles the empty data case?

rufman committed:
fix(a-msgpack): use data.length instead of flag

Instead of using a flag to only write the header, just return an empty
Uint8 array to write only the header without any data, for an extension
type.

timoha commented on packages/a-msgpack/src/Encoder.ts:
```diff
@@ -359,6 +359,11 @@ export class Encoder {
   }
 
   encodeExtension(ext: ExtData): void {
```
looks like you don't need to split this function anymore. it's up to you if you want to revert it

timoha commented on packages/a-msgpack/src/ExtensionCodec.ts:
```diff
@@ -3,46 +3,54 @@
+    let type = null;
+    for (let i = 0; i < this.identifiers.length; i++) {
+      const id = this.identifiers[i];
+      if (id(object)) {
```
wondering if you could have a map where the key is the typeof Wildcard/Pointer and then you'd encoder/decoder by using `typeof` function?

rufman replied:
instanceof might work, but there could be a case where it doesn't. Like when it's used across threads. Because the instance could be different.

For now I guess I can do that and worry about that later

timoha replied:
didn't know class can have different identity in diff threads 🤦
you could alternatively use `.type` field maybe as you seem to require it everywhere?

rufman committed:
fix(a-msgpack): revert extension rewrite

rufman committed:
Merge branch 'trunk' into feature/wildcard
